lightning/src/routing/scoring.rs

```diff
@@ -460,25 +458,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	/// a payment.
+	///
+	/// The success probability is determined by the effective channel capacity, the payment amount,
+	/// and knowledge learned from prior successful and unsuccessful payments. The knowledge learned
+	/// is reduced over time based on [`liquidity_offset_half_life`].
+	///
+	/// Default value: 1,000 msat
+	///
+	/// [`liquidity_offset_half_life`]: Self::liquidity_offset_half_life
+	pub liquidity_penalty_multiplier_msat: u64,
+
+	/// The time required to elapse before any knowledge learned about channel liquidity balances is
+	/// cut in half.
+	///
+	/// The bounds are defined in terms offsets and are initially zero. Increasing the offsets gives
+	/// tighter bounds on the channel liquidity balance. Thus, halving the offsets decreases the
+	/// certainty of the channel liquidity balance.
+	///
+	/// Default value: 1 hour
+	///
+	/// # Note
+	///
+	/// When built with the `no-std` feature, time will never elapse. Therefore, the channel
+	/// liquidity knowledge will never decay except when the bounds cross.
+	pub liquidity_offset_half_life: Duration,
+}
+
+impl_writeable_tlv_based!(ProbabilisticScoringParameters, {
+	(0, liquidity_penalty_multiplier_msat, required),
+	(2, liquidity_offset_half_life, required),
+});
+
+/// Accounting for channel liquidity balance uncertainty.
+///
+/// Direction is defined in terms of [`NodeId`] partial ordering, where the source node is the
+/// first node in the ordering of the channel's counterparties. Thus, swapping the two liquidity
+/// offset fields gives the opposite direction.
+struct ChannelLiquidity<T: Time> {
+	min_liquidity_offset_msat: u64,
+	max_liquidity_offset_msat: u64,
+	last_updated: T,
+}
+
+/// A view of [`ChannelLiquidity`] in one direction assuming a certain channel capacity and decayed
+/// with a given half life.
+struct DirectedChannelLiquidity<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> {
+	min_liquidity_offset_msat: L,
+	max_liquidity_offset_msat: L,
+	capacity_msat: u64,
+	last_updated: U,
+	now: T,
+	half_life: Duration,
+}
+
+/// The likelihood of an event occurring.
+enum Probability {
+	Zero,
+	One,
+	Ratio { numerator: u64, denominator: u64 },
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> ProbabilisticScorerUsingTime<G, T> {
+	/// Creates a new scorer using the given scoring parameters for sending payments from a node
+	/// through a network graph.
+	pub fn new(
+		params: ProbabilisticScoringParameters, node_pubkey: &PublicKey, network_graph: G
+	) -> Self {
+		Self {
+			params,
+			node_id: NodeId::from_pubkey(node_pubkey),
+			network_graph,
+			channel_liquidities: HashMap::new(),
+		}
+	}
+
+	#[cfg(test)]
+	fn with_channel(mut self, short_channel_id: u64, liquidity: ChannelLiquidity<T>) -> Self {
+		assert!(self.channel_liquidities.insert(short_channel_id, liquidity).is_none());
+		self
+	}
+}
+
+impl Default for ProbabilisticScoringParameters {
+	fn default() -> Self {
+		Self {
+			liquidity_penalty_multiplier_msat: 1000,
```

Comment:
This seems only relevent if (as one should) also fees are taken into consideration as a feature to compute the cost of a hop / path (I think you internally name the cost Score, so please don't get confused when I talk about cost instead of score). As our benchmarks and prior to them simulations in c-lightning demonstrated (https://github.com/ElementsProject/lightning/pull/4771#issuecomment-930173831) using the harmonic mean instead of a weighted arithmetic mean to combine several features (negative log probs, fees, cltv delta) worked better. In case you would do the same this value seems useless. Otherwise one could of course ask why is it chosen as `1000` right now.

Reply:
What do you mean "better", here? There is no objective metric by which one could argue one way of combining different goals is "better" than another - hence why this is configurable.

Reply:
> What do you mean "better", here? There is no objective metric by which one could argue one way of combining different goals is "better" than another - hence why this is configurable.

Yes you have already mentioned that on the mailing list and I agree with you (never disagreed) that there is not one single objective, universally true metric here. The way I used the word "better" was having in mind to combine the relevant features which had previously been identified (eg probabilities or fees)

That being said the topic here goes far beyond Lightning Path finding / payment delivery. In many cases where you have several features (like in our case Probability to model reliability (and with it potentially speed), fees to model actual costs, latency or geodistance to model speed, ...) you will always have the questions: "**How should those features be combined into one single score / metric?**". It turns out that many people intuitively take the arithmetic mean which is essentially just adding the features or a weighted arithmetic mean which is almost the same as normalizing the features and adding them without weight. You are currently also doing this and I would strongly advice against doing that.

You take your old score and add the new one and you currently weight it with a factor of 1000 as in `liquidity_penalty_multiplier_msat`. In almost all situations in my life when I was doing data science / machine learning it turned out that in practice the arithmetic mean `a+w*b` is not as useful as using `a*b/(a+b)` (I ignore constants). The later seems to have several advantages in practise: 

1. You do not need to learn and test for the weight(s)
2. features (in our case for channels) might not be correlated. so a weighting for one channel might be totally wrong for another channel
3. It seems to be more resilient against outliers and seems to be able to capture the mix of features better even if they are in different orders of magnitude or are differently distributed. 

I have [discussed this in great length with the c-lightning team](https://github.com/ElementsProject/lightning/pull/4771#issuecomment-925990761) - in particular @cdecker - and [my preliminary simulations](https://github.com/ElementsProject/lightning/pull/4771#issuecomment-930173831) as well as [our mainnet experiments and evaluations show that arithmetic mean heavily underperformed in comparison to the other way to compute a mean](https://medium.com/blockstream/c-lightning-v0-10-2-bitcoin-dust-consensus-rule-33e777d58657). Also recently I was made aware that there might actually be deeper mathematical motivation in there which in the field of economics and risk management is being used all the time and might have already been [discovered in the early 18th hundred by Bernoulli in this work](https://www.zoology.ubc.ca/let/pdfs/Stearns2000.pdf) - though I haven't looked deeper into this yet.

Also on a side note. The arithmetic mean obviously has the advantage of being linear and useful for the case of min cost flows if all the features are linear and convex. This is not true for the mean that I suggest. But as your code currently only looks at dijkstra we don't need to have a linear or convex cost function.

Also as suggested to the c-lightning team. if you don't believe me it should be very easy to have both versions of mean computation and do a simple A/B test against any of your favourite test criteria (I usually like "number of attempts" or time to delivery) to see if you want to keep the arithmetic mean here

Reply:
I'm still very confused by this, I read through the links you provide and none of them discuss failure modes or seem to provide any justification for harmonic mean here, aside from some anecdotal suggestions. The discussion from c-lightning just links to an lnd issue, which provides no justification, and it looks like c-lightning stuck with simply adding the scores as well?

Indeed, the code here (and the current model) focuses entirely on *fees* as the metric we're optimizing for. As a result of this, we denominate the scorer as "amount of fee you're willing to pay to *avoid* this channel". This is not only nicely simple from a theoretical perspective, but also simple enough that our users can at least roughly understand the metric and can set the tunables here without much interaction from us - they can *easily* run trivial calculations to determine how much their "avoid channels" cost will be, and can tune things appropriately based on how much fee they're willing to pay. In the future we could even allow these tunables to be set on a per-payment basis, given we have a model here that doesn't incorporate the concrete fee values.

Indeed, in the future, I'd very much like to have multiple top-to-bottom routing algorithms, eg there has been very little work done on how you'd design a lightning routing algorithm "for privacy", without seriously considering fees or time-to-pay/-retries, but even just as-is, users having the ability to drastically tune the knob between retry-count and fees-paid seems pretty important, which gets much harder with the harmonic mean, IIUC.

Reply:
https://github.com/ElementsProject/lightning/blob/a98ccac7774297f255f0a4f585384609af87c3d6/plugins/libplugin-pay.c#L735 

shows what c-lightning does and it was evaluated bu conducting mainnet evaluation: https://medium.com/blockstream/c-lightning-v0-10-2-bitcoin-dust-consensus-rule-33e777d58657

The blog article however left out the comparison between the means but I remember evaluating the statistics for test payments myself. 

Of course the evaluation that @cdecker and I did might not have been sufficiently large or realistic from the assumptions.

Reply:
btw I think the much bigger argument for using an arithmetic mean is that it is linear which means that if the features are linear the cost function stays linear (of if features are convex the cost function stays convex). When using min cost flow solvers to compute optimal splits (given the cost function and assuming the cost function properly captures the goals of the node) then the arithmetic mean will not have a negative impact on the runtime of the min cost flows solvers (assuming all features are at worst convex). As the cost function is currently only used in dijkstra we need neither linear nor convex so we don't care.